### DISCUSSION TITLE
Preloading set of images per trial

### DISCUSSION THREAD

**Discussion by NadineDijkstra** [jspsych/jsPsych#724] [Q&A] [answered]
Hi hi, I am setting up an experiment in which I want to show a set of 20 images for 100 ms each - the images are different every trial. I am using the jspsych-animation plugin which works like a charm but sometimes the server is too slow and it shows an image-pictogram instead of one of the images. I know I could preload all images at the start of the experiment with the jspsych-init function but in my case that would be way too many. Is there a way to preload a set of images prior to every trial (e.g. in the ITI) and then remove them again? Thanks!

**Comment by vijaymarupudi** [comment, marked as answer]
Before a trial with images, or a block of trials, you can use the [call-function](https://www.jspsych.org/plugins/jspsych-call-function/) plugin to preload the images. I'm using [getDisplayElement()](https://www.jspsych.org/core_library/jspsych-core/#jspsychgetdisplayelement) and [pluginAPI.preloadImages()](https://www.jspsych.org/core_library/jspsych-pluginAPI/#jspsychpluginapipreloadimages)

```javascript
var images = ["image1.png", "image2.png"]

var trial = {
  type: "call-function",
  async: true,
  func: function(done) {
    jsPsych.getDisplayElement().innerHTML = '<p>Loading... Please wait.</p>'
    jsPsych.pluginAPI.preloadImages(images, function() {
      done({ preload: "success" });
    })
  }
};
```

**Reply by jodeleeuw** [reply]
This is exactly what I would recommend. You also asked about removing the images. The browser will take care of that for you if the cache grows too large, so you don't need to worry about managing that side of it.

**Reply by NadineDijkstra** [reply]
Hm, I can't seem to get it to work. Here is my code - can you see where I went wrong? Thanks!


var preload_images = {
      type: "call-function",
      async: true,
      func: function(done) {
        jsPsych.getDisplayElement().innerHTML = "Loading... Please wait."
        jsPsych.pluginAPI.preloadImages(jsPsych.timelineVariable('stimulus'), function() {
          done({ preload: "success" });
        })
      }
    };

    var show_dynamic_stimulus = {
      type: 'animation',
      stimuli: jsPsych.timelineVariable('stimulus'),
      choices: jsPsych.NO_KEYS,
      frame_isi: 0,
      frame_time: Math.round(stim_dur/num_steps)
    };

    var response_screen = {
      type: 'html-keyboard-response',
      stimulus: 'Was there a stimulus on the screen?',
      choices: ['f', 'j'],
      prompt: 'Yes [F] or no [J]',
      data: jsPsych.timelineVariable('data'),
      on_finish: function(data){
        data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
      },
    };

    var detection_trial_procedure = {
      timeline: [preload_images,show_dynamic_stimulus,response_screen],
      timeline_variables: dynamic_stimulus_array,
      sample: {
        type: 'fixed-repetitions',
        size: 5,
      }
    }
    timeline.push(detection_trial_procedure);

**Reply by vijaymarupudi** [reply]
When you use timeline variables in a function, the second parameter should be true.

See second example under [here](https://www.jspsych.org/overview/timeline/#timeline-variables).

**Reply by NadineDijkstra** [reply]
Thank you that helps! It now works most times but every now and then, seemingly random, I get a message that says 'experiment failed to load'. Can you think of what would be going wrong there?

**Reply by NadineDijkstra** [reply]
Also, I just saw that the animation plug-in that I am using has the following code:

```
jsPsych.plugins.animation = (function() {

  var plugin = {};

  jsPsych.pluginAPI.registerPreload('animation', 'stimuli', 'image');
```

Doesn't that mean that the stimuli should be preloaded automatically when I run this function?

**Reply by vijaymarupudi** [reply]
Do you see any errors in the console when "experiment fails to load" shows up?

That function registers it to preload at the beginning of the experiment, it doesn't preload immediately.

**Reply by NadineDijkstra** [reply]
Ah I wasn't aware of the console error reporting :') This helps a lot, thanks, it's working now!
